src/stdune/set.ml

```diff
@@ -1,49 +1,71 @@
 module type S = Set_intf.S
 
 module Make(Elt : Comparable.S) : S with type elt = Elt.t = struct
-  module M = MoreLabels.Set.Make(struct
-      type t = Elt.t
-      let compare a b = Ordering.to_int (Elt.compare a b)
-    end)
+  module M = Map.Make(Elt)
 
   include struct
     [@@@warning "-32"]
     (* [map] is only available since 4.04 *)
     let map ~f t =
-      M.elements t
+      M.keys t
       |> List.map ~f
       |> M.of_list
 
     (* Since 4.05 *)
     let to_opt f t =
       match f t with
-      | x -> Some x
-      | exception Not_found -> None
+      | Some (x, _) -> Some x
+      | None -> None
     let choose_opt  = to_opt M.choose
-    let min_elt_opt = to_opt M.min_elt
-    let max_elt_opt = to_opt M.max_elt
+    let min_elt_opt = to_opt M.min_binding
+    let max_elt_opt = to_opt M.max_binding
   end
 
-  include M
-
-  let to_list = elements
-
-  let mem t x = mem x t
-  let add t x = add x t
-  let remove t x = remove x t
-  let compare a b = Ordering.of_int (compare a b)
-  let is_subset t ~of_ = subset t of_
-  let iter t ~f = iter t ~f
-  let map t ~f = map t ~f
-  let fold t ~init ~f = fold t ~init ~f
-  let for_all t ~f = for_all t ~f
-  let exists t ~f = exists t ~f
-  let filter t ~f = filter t ~f
-  let partition t ~f = partition t ~f
+  type elt = Elt.t
+
+  type t = unit M.t
+
+  let to_list = M.keys
+
+  let is_empty = M.is_empty
+  let cardinal = M.cardinal
+  let equal = M.equal ~equal:(fun () () -> true)
+  let empty = M.empty
+  let singleton k = M.singleton k ()
+
+  let diff = M.merge ~f:(fun _ x y ->
```

Comment:
These implementations of `diff` and `inter` are pretty inefficient. They're linear-time in the size of the larger map (in the worst and in the best case), whereas the old functions are linear-time in the size of the smaller map (or perhaps `O(m*log(n/m))`?) in the worst case and faster than that in the best case (when there's some clustering of keys).

I can't see enough public things exposed in `Map` to implement equally efficient versions though.

Not sure what we should do about this.

Reply:
Hmm. I believe the original idea was from @bobot. Let's see if has a solution for this

Reply:
Another options is to vendor the map implementation from base or the stdlib. That shouldn't be so huge on its own, right?

Reply:
I guess that can work, yeah. If we're doing a major change like that, I wish we used the more flexible types from Base which let you write functions generic over all maps. 

That said, base doesn't expose the efficient hedge-union operations either. We'd have to improve that.

Reply:
> I guess that can work, yeah. If we're doing a major change like that, I wish we used the more flexible types from Base which let you write functions generic over all maps.

We should do that too. I believe we can do that separately however. We can introduce a polymorphic module for maps (like you did for hashtable) and we can slowly start porting things. Porting everything at once seems too hard.

Reply:
> Porting everything at once seems too hard.

Oh, I thought Base.Map.Make existed which would give us a smooth way to migrate. but I guess I was thinking of `core_kernel`.